### PR TITLE
service: Include TS code module dir in `.env` search

### PR DESCRIPTION
**Pull request by bkeryan**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-python/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?

Update `.env` search to include some fallback options if the current directory and its parents don't contain a `.env` file:
- For PyInstaller, use the EXE path
- For scripts that have a `__main__` module that exists on disk, use the script path
- For TestStand code modules (and possibly other cases), use the path of the topmost caller that isn't part of `ni_measurementlink_service`

### Why should this Pull Request be merged?

Fixes [AB#2565326](https://dev.azure.com/ni/94b22d7b-ad7b-4f5e-88f0-867910f91c94/_workitems/edit/2565326)

### What testing has been done?

Ran tests.
Manually tested with nidcpower_source_dc_voltage TestStand sequence.

### CODE REVIEW

**Comment by bkeryan** [comment]
@mshafer-NI This implements some of the ideas we discussed earlier.

File: tests/unit/test_dotenvpath.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+import pytest
+
+import ni_measurementlink_service
+from ni_measurementlink_service import _dotenvpath
+
+
+@pytest.mark.parametrize("dotenv_exists", [False, True])
+def test___dotenv_exists_varies___has_dotenv_file___matches_dotenv_exists(
+    dotenv_exists: bool, tmp_path: Path
```
Comment (dixonjoel) [inline tests/unit/test_dotenvpath.py:11]: How does `tmp_path` get filled in?

**Review by dixonjoel — APPROVED** [review]

File: tests/unit/test_dotenvpath.py (hunk 1 above)
Comment (bkeryan) [reply tests/unit/test_dotenvpath.py:11]: It's a standard fixture included with pytest. 
https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html

File: tests/unit/test_dotenvpath.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+def test___get_nims_path___returns_nims_path() -> None:
+    assert _dotenvpath._get_nims_path() == Path(ni_measurementlink_service.__file__)
+
+
+def test___get_script_or_exe_path___returns_pytest_path() -> None:
+    path = _dotenvpath._get_script_or_exe_path()
+
+    assert path is not None
+    assert path.parent.name in ["pytest", "pytest.exe"]
+    assert path.name == "__main__.py"
```
Comment (bkeryan) [inline tests/unit/test_dotenvpath.py:36]: This is too specific.

Linux:
```
FAILED tests/unit/test_dotenvpath.py::test___get_script_or_exe_path___returns_pytest_path - AssertionError: assert 'bin' in ['pytest', 'pytest.exe']
 +  where 'bin' = PosixPath('/home/runner/work/measurementlink-python/measurementlink-python/.venv/bin').name
 +    where PosixPath('/home/runner/work/measurementlink-python/measurementlink-python/.venv/bin') = PosixPath('/home/runner/work/measurementlink-python/measurementlink-python/.venv/bin/pytest').parent
```

Windows with Python 3.11:
```
FAILED tests/unit/test_dotenvpath.py::test___get_caller_path___returns_this_modules_path - AssertionError: assert WindowsPath('D:/a/measurementlink-python/measurementlink-python/.venv/Lib/site-packages/pluggy/_callers.py') == WindowsPath('D:/a/measurementlink-python/measurementlink-python/tests/unit/test_dotenvpath.py')
 +  where WindowsPath('D:/a/measurementlink-python/measurementlink-python/.venv/Lib/site-packages/pluggy/_callers.py') = <function _get_caller_path at 0x00000268B012C0E0>()
 +    where <function _get_caller_path at 0x00000268B012C0E0> = _dotenvpath._get_caller_path
 +  and   WindowsPath('D:/a/measurementlink-python/measurementlink-python/tests/unit/test_dotenvpath.py') = Path('D:\\a\\measurementlink-python\\measurementlink-python\\tests\\unit\\test_dotenvpath.py')
```

File: ni_measurementlink_service/_dotenvpath.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    return None
+
+
+def _get_caller_path() -> Optional[Path]:
+    """Get the path of the module calling into ni_measurementlink_service, if possible."""
+    nims_path = _get_nims_path()
+    for frame, _ in traceback.walk_stack(None):
+        if frame.f_code.co_filename:
+            module_path = Path(frame.f_code.co_filename)
+            if module_path.exists() and not module_path.is_relative_to(nims_path):
```
Comment (bkeryan) [inline ni_measurementlink_service/_dotenvpath.py:46]: `is_relative_to` isn't supported in Python 3.8.
```
FAILED tests/unit/test_dotenvpath.py::test___get_caller_path___returns_this_modules_path - AttributeError: 'WindowsPath' object has no attribute 'is_relative_to'
```

File: tests/unit/test_dotenvpath.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    subdirs = [tmp_path / "a", tmp_path / "a" / "b", tmp_path / "a" / "b" / "c"]
+    for dir in subdirs:
+        dir.mkdir()
+
+    assert _dotenvpath._has_dotenv_file(tmp_path) == dotenv_exists
+    assert all([_dotenvpath._has_dotenv_file(p) == dotenv_exists for p in subdirs])
+
+
+def test___get_caller_path___returns_this_modules_path() -> None:
+    assert _dotenvpath._get_caller_path() == Path(__file__)
```
Comment (bkeryan) [inline tests/unit/test_dotenvpath.py:24]: Windows with Python 3.11:
```
FAILED tests/unit/test_dotenvpath.py::test___get_caller_path___returns_this_modules_path - AssertionError: assert WindowsPath('D:/a/measurementlink-python/measurementlink-python/.venv/Lib/site-packages/pluggy/_callers.py') == WindowsPath('D:/a/measurementlink-python/measurementlink-python/tests/unit/test_dotenvpath.py')
 +  where WindowsPath('D:/a/measurementlink-python/measurementlink-python/.venv/Lib/site-packages/pluggy/_callers.py') = <function _get_caller_path at 0x00000268B012C0E0>()
 +    where <function _get_caller_path at 0x00000268B012C0E0> = _dotenvpath._get_caller_path
 +  and   WindowsPath('D:/a/measurementlink-python/measurementlink-python/tests/unit/test_dotenvpath.py') = Path('D:\\a\\measurementlink-python\\measurementlink-python\\tests\\unit\\test_dotenvpath.py')
```

Comment (bkeryan) [reply tests/unit/test_dotenvpath.py:24]: In Python 3.11 and later, `traceback.walk_stack(None)` skips more stack frames than in the past: https://github.com/python/cpython/issues/96092

**Pull request merged**